bokehjs/src/lib/models/widgets/file_input.ts

```diff
@@ -16,43 +16,69 @@ export class FileInputView extends WidgetView {
     if (this.dialogEl == null) {
       this.dialogEl = document.createElement('input')
       this.dialogEl.type = "file"
-      this.dialogEl.multiple = false
-      this.dialogEl.onchange = (e) => this.load_file(e)
+      this.dialogEl.multiple = this.model.multiple
+      this.dialogEl.onchange = () => {
+        const {files} = this.dialogEl
+        if (files != null) {
+          this.load_files(files)
+        }
+      }
       this.el.appendChild(this.dialogEl)
     }
     if (this.model.accept != null && this.model.accept != '')
       this.dialogEl.accept = this.model.accept
+
     this.dialogEl.style.width = `{this.model.width}px`
     this.dialogEl.disabled = this.model.disabled
   }
 
-  load_file(e: any): void {
-    const reader = new FileReader()
-    this.model.filename = e.target.files[0].name
-    reader.onload = (e) => this.file(e)
-    reader.readAsDataURL(e.target.files[0])
+  async load_files(files: FileList): Promise<void> {
+    const value: string[] = []
+    const filename: string[] = []
+    const mime_type: string[] = []
+    let i: number
+
+    for (i = 0; i < files.length; i++){
+      filename.push(files[i].name)
+      const data_url = await this.readfile(files[i])
+      const [, mime, , data] = data_url.split(/[:;,]/, 4)
+      value.push(data)
+      mime_type.push(mime)
+    }
+    if (this.model.multiple) {
+      this.model.filename = filename
+      this.model.mime_type = mime_type
+      this.model.value = value
+    } else {
+      this.model.filename = filename[0]
+      this.model.mime_type = mime_type[0]
+      this.model.value = value[0]
+    }
   }
 
-  file(e: any): void {
-    const file = e.target.result
-    const file_arr = file.split(",")
-
-    const content = file_arr[1]
-    const header = file_arr[0].split(":")[1].split(";")[0]
-
-    this.model.value = content
-    this.model.mime_type = header
-
+  readfile(file: File): Promise<string> {
+    return new Promise<string>((resolve) => {
+      const reader = new FileReader()
+      reader.onload = (e) => {
+        if (e.target) {
+          resolve(e.target.result ? String(e.target.result) : "")
+        } else {
+          resolve("")
+        }
```

Comment:
This block of code still needs some refinement. In general, `e.target.result` can be a `string` or an `ArrayBuffer` at `String()` call. However, `String()` applied to an `ArrayBuffer` gives a result far from what we would expected:
```ts
$ node
Welcome to Node.js v12.14.1.
Type ".help" for more information.
> x = new ArrayBuffer(10)
ArrayBuffer {
  [Uint8Contents]: <00 00 00 00 00 00 00 00 00 00>,
  byteLength: 10
}
> String(x)
'[object ArrayBuffer]'
```
However, we don't have to deal with `ArrayBuffer` at all, because using `readAsDataURL()` guarantees that `result` will be only of type `string | null`. Unfortunately encoding this at type-level is hard or perhaps not possible at all. So, my suggestion is to cast `result as string` to help the type system and not use otherwise incorrectly behaving `String()`.

Additionally, returning an empty string on failure will result in a failure on new line 44, where we assume a data URL. The empty string on line 66 can be eliminated easily, because we can assume `e.target == reader` and remove the if statement altogether. Or alternatively we could use `e.target!` type-level construct, telling the type-checker that we know `null` can't appear, but we can't express this by better and safer means. The remaining question is whether `result` can really be `null` at this point. After consulting FileReader's specification, I'm not sure what to expect. So, we have two options. Either carry this `null` around, which will likely mean that any errors will be silently ignored. Alternatively, we can reject the promise with an error. This is the solution I'm leaning on. 

In summary I would proceed with the following code:
```ts
    return new Promise<string>((resolve, reject) => {
      const reader = new FileReader()
      reader.onload = () => {
        const {result} = reader
        if (result != null) {
          resolve(result as string)
        } else {
          reject(reader.error ?? new Error(`unable to read '${file.name}'`))
        }
      }
      reader.readAsDataURL(file)
    })
```

Reply:
Thank you!
I implemented this in a further commit (copy/paste because such refinements are beyond my skills). 
I tested to trigger a reject() by selecting/uploading a dead soft-link without a target-file but this resulted merely in "no file selected". Selecting three files with one being the dead soft link resulted in "two files selected". So no error under this condition. (While the shell informs "no such file or directory" if I want to "> less dead_link.txt".)